### PR TITLE
Batch transaction

### PR DESCRIPTION
**Pull request by davimacedo**
This PR empowers the current Parse Server's `/batch/` endpoint with transaction capability for both MongoDB (>4.0) and Postgres. You can find an example of request below:

```
        request({
          method: 'POST',
          headers: headers,
          url: 'http://localhost:8378/1/batch',
          body: JSON.stringify({
            requests: [
              {
                method: 'POST',
                path: '/1/classes/MyObject',
                body: { key: 'value1' },
              },
              {
                method: 'POST',
                path: '/1/classes/MyObject',
                body: { key: 'value2' },
              },
            ],
            transaction: true,
          }),
        })
```

Close https://github.com/parse-community/parse-server/issues/2159

### CODE REVIEW

**Comment by davimacedo** [comment]
@dplewis sorry for flooding you with so many commits in this PR. It is actually working, but MongoDB only supports transactions when running version >4 in a wiredTiger replicaset. I am trying to figure out how to setup this environment in Travis. I will let you know then.

**Comment by dplewis** [comment]
No problem take your time. I know this is a highly requested feature.

File: src/batch.js [hunk 1]
```diff
@@ -83,30 +83,57 @@ function handleBatch(router, req) {
...
-    // Construct a request that we can send to a handler
-    const request = {
-      body: restRequest.body,
-      config: req.config,
-      auth: req.auth,
-      info: req.info,
-    };
+  let initialPromise = Promise.resolve();
+  if (req.body.transaction === true) {
+    initialPromise = req.config.database.createTransactionalSession();
```
Comment (dplewis) [inline src/batch.js]: Good job getting this one in. I can't wait to use it.

Can you add this to direct access?

https://github.com/parse-community/parse-server/blob/master/src/ParseServerRESTController.js

Comment (davimacedo) [reply src/batch.js]: done

**Comment by andreasanta** [comment]
Guys this is great work! Any plans to make it accessible from the JS SDK and the remaning SDKs soon? Can I help with that?

**Comment by davimacedo** [comment]
Yes. There is an open PR (https://github.com/parse-community/Parse-SDK-JS/pull/1090) that we are working on.

**Pull request merged**